File: sdks/dotnet/Sequins.OpenTelemetry/SequinsHandle.cs
using OpenTelemetry;
using OpenTelemetry.Logs;
using OpenTelemetry.Metrics;
using OpenTelemetry.Trace;

namespace Sequins.OpenTelemetry;

/// <summary>
/// Handle returned by <see cref="SequinsExtensions.AddSequins"/> or <see cref="SequinsBuilder.Build"/>.
/// Provides access to all configured OpenTelemetry providers.
/// </summary>
public sealed class SequinsHandle : IDisposable
{
    /// <summary>Gets the configured TracerProvider.</summary>
    public TracerProvider TracerProvider { get; }

    /// <summary>Gets the configured MeterProvider.</summary>
    public MeterProvider MeterProvider { get; }

    /// <summary>Gets the configured LoggerProvider.</summary>
    public LoggerProvider LoggerProvider { get; }

    internal SequinsHandle(TracerProvider tracerProvider, MeterProvider meterProvider, LoggerProvider loggerProvider)
    {
        TracerProvider = tracerProvider;
        MeterProvider = meterProvider;
        LoggerProvider = loggerProvider;
    }

    /// <inheritdoc />
    public void Dispose()
    {
        TracerProvider.Dispose();
        MeterProvider.Dispose();
        LoggerProvider.Dispose();
    }
}

File: sdks/php/src/SequinsHandle.php
<?php

declare(strict_types=1);

namespace Sequins\Otel;

use OpenTelemetry\API\Logs\LoggerProviderInterface;
use OpenTelemetry\API\Metrics\MeterProviderInterface;
use OpenTelemetry\API\Trace\TracerProviderInterface;
use OpenTelemetry\SDK\Logs\LoggerProvider;
use OpenTelemetry\SDK\Metrics\MeterProvider;
use OpenTelemetry\SDK\Trace\TracerProvider;

/**
 * Handle returned by Sequins::init(). Provides access to all configured OpenTelemetry providers.
 */
final class SequinsHandle
{
    public function __construct(
        private readonly TracerProviderInterface $tracerProvider,
        private readonly MeterProviderInterface $meterProvider,
        private readonly LoggerProviderInterface $loggerProvider,
    ) {
    }

    /**
     * Returns the configured TracerProvider.
     * Use getTracerProvider()->getTracer('name') to create tracers.
     */
    public function getTracerProvider(): TracerProviderInterface
    {
        return $this->tracerProvider;
    }

    /**
     * Returns the configured MeterProvider.
     * Use getMeterProvider()->getMeter('name') to create meters.
     */
    public function getMeterProvider(): MeterProviderInterface
    {
        return $this->meterProvider;
    }

    /**
     * Returns the configured LoggerProvider.
     */
    public function getLoggerProvider(): LoggerProviderInterface
    {
        return $this->loggerProvider;
    }

    /**
     * Flushes all pending telemetry and shuts down all providers.
     * This is registered automatically as a shutdown function.
     */
    public function shutdown(): void
    {
        if ($this->tracerProvider instanceof TracerProvider) {
            $this->tracerProvider->shutdown();
        }
        if ($this->meterProvider instanceof MeterProvider) {
            $this->meterProvider->shutdown();
        }
        if ($this->loggerProvider instanceof LoggerProvider) {
            $this->loggerProvider->shutdown();
        }
    }
}
